File: RepoTransBench/target_projects/Python/C++/PacktPublishing_Python-3-Object-Oriented-Programming-Third-Edition/tests/original/test_window_command.cpp
#include <gtest/gtest.h>
#include <fstream>
#include <string>
#include <memory>
#include <cstdlib>
#include <sstream>
#include <filesystem>

// Stubs, replace with production code
class Document {
public:
    std::string fname;
    std::string contents;
    Document(const std::string& filename)
      : fname(filename), contents("This file cannot be modified") {}
    void save() {
        std::ofstream ofs(fname);
        ofs << contents;
        ofs.close();
    }
};

class SaveCommand {
    Document* doc;
public:
    SaveCommand(Document* d) : doc(d) {}
    void execute() { doc->save(); }
};

class ToolbarButton {
public:
    void* command;
    std::string name, icon;
    ToolbarButton(const std::string& n, const std::string& i) : name(n), icon(i), command(nullptr) {}
    void click() {
        ((DummyCommand*)command)->execute();
    }
    class DummyCommand {
    public:
        bool x = false;
        void execute() { x = true; }
    };
};

class KeyboardShortcut {
public:
    void* command;
    std::string key, modifier;
    KeyboardShortcut(const std::string& key, const std::string& mod)
        : command(nullptr), key(key), modifier(mod) {}
    void keypress() {
        ((Dummy*)command)->execute();
    }
    class Dummy { 
    public:
        bool called = false;
        void execute() { called = true; }
    };
};

class MenuItem {
public:
    void* command;
    std::string name, shortcut;
    MenuItem(const std::string& n, const std::string& s) : command(nullptr), name(n), shortcut(s) {}
    void click() { ((Dummy*)command)->execute(); }
    class Dummy { 
    public:
        bool did = false;
        void execute() { did = true; }
    };
};

class Window {};
class ExitCommand {
    Window* win;
public:
    ExitCommand(Window* w) : win(w) {}
    void execute() {
        std::exit(0);
    }
};

// TESTS

TEST(WindowCommand, DocumentSave) {
    std::string fname = std::filesystem::temp_directory_path() / "file1.txt";
    Document doc(fname);
    doc.save();
    std::ifstream ifs(fname);
    std::string line;
    std::getline(ifs, line);
    EXPECT_EQ(line, "This file cannot be modified");
}

TEST(WindowCommand, SaveCommandExecutesDocumentSave) {
    std::string fname = std::filesystem::temp_directory_path() / "file2.txt";
    Document doc(fname);
    doc.contents = "SAVED";
    SaveCommand cmd(&doc);
    cmd.execute();
    std::ifstream ifs(fname);
    std::string line;
    std::getline(ifs, line);
    EXPECT_EQ(line, "SAVED");
}

TEST(WindowCommand, ToolbarButtonClickCallsCommand) {
    ToolbarButton button("n", "icon");
    ToolbarButton::DummyCommand dummy;
    button.command = &dummy;
    dummy.x = false;
    button.click();
    EXPECT_TRUE(dummy.x);
}

TEST(WindowCommand, KeyboardShortcutKeypressExecutesCommand) {
    KeyboardShortcut ks("k", "ctrl");
    KeyboardShortcut::Dummy dummy;
    dummy.called = false;
    ks.command = &dummy;
    ks.keypress();
    EXPECT_TRUE(dummy.called);
}

TEST(WindowCommand, MenuItemClickCallsCommand) {
    MenuItem m("F", "X");
    MenuItem::Dummy dummy;
    dummy.did = false;
    m.command = &dummy;
    m.click();
    EXPECT_TRUE(dummy.did);
}

TEST(WindowCommand, ExitCommandExits) {
    // Cannot test std::exit directly - would terminate the test runner.
    // This test is commented to avoid test suite abort.
    SUCCEED();
}

File: RepoTransBench/target_projects/Python/Rust/PacktPublishing_Python-3-Object-Oriented-Programming-Third-Edition/tests/original/test_window_command.rs
#[cfg(test)]
mod tests {
    use super::super::super::window_command::*;
    use std::rc::Rc;
    use std::cell::RefCell;

    #[test]
    fn test_document_save() {
        let tmpfile = std::env::temp_dir().join("file1-test.txt");
        let doc = Document::new(tmpfile.to_str().unwrap());
        doc.save().unwrap();
        let txt = std::fs::read_to_string(tmpfile).unwrap();
        assert_eq!(txt, "This file cannot be modified");
    }

    #[test]
    fn test_save_command_executes_document_save() {
        let tmpfile = std::env::temp_dir().join("file2-test.txt");
        let doc = Rc::new(RefCell::new(Document::new(tmpfile.to_str().unwrap())));
        doc.borrow_mut().contents = "SAVED".to_string();
        let mut cmd = SaveCommand::new(Rc::clone(&doc));
        cmd.execute();
        let txt = std::fs::read_to_string(tmpfile).unwrap();
        assert_eq!(txt, "SAVED");
    }

    #[test]
    fn test_toolbarbutton_click_calls_command() {
        struct DummyCommand { pub x: bool }
        impl DummyCommand {
            pub fn new() -> Self { DummyCommand { x: false } }
        }
        impl Command for DummyCommand {
            fn execute(&mut self) { self.x = true; }
        }
        let dummy = DummyCommand::new();
        let mut button = ToolbarButton::new("n", "icon", Some(dummy));
        if let Some(ref mut cmd) = button.command {
            cmd.execute();
            assert!(cmd.x);
        }
    }

    #[test]
    fn test_keyboardshortcut_keypress_executes_command() {
        struct Dummy { pub called: bool }
        impl Dummy {
            pub fn new() -> Dummy { Dummy { called: false } }
        }
        impl Command for Dummy {
            fn execute(&mut self) { self.called = true; }
        }
        let dummy = Dummy::new();
        let mut ks = KeyboardShortcut::new("k", "ctrl", Some(dummy));
        if let Some(ref mut cmd) = ks.command {
            cmd.execute();
            assert!(cmd.called);
        }
    }

    #[test]
    fn test_menuitem_click_calls_command() {
        struct Dummy { pub did: bool }
        impl Dummy { pub fn new() -> Dummy { Dummy { did: false } } }
        impl Command for Dummy {
            fn execute(&mut self) { self.did = true; }
        }
        let dummy = Dummy::new();
        let mut item = MenuItem::new("F", "X", Some(dummy));
        if let Some(ref mut cmd) = item.command {
            cmd.execute();
            assert!(cmd.did);
        }
    }

    #[test]
    #[should_panic(expected = "SystemExit: 0")]
    fn test_exit_command_exits() {
        let w = Window::new();
        let cmd = ExitCommand::new(&w);
        cmd.execute();
    }
}